lib/faktory_worker/connection.ex

```diff
@@ -3,17 +3,60 @@ defmodule FaktoryWorker.Connection do
 
   """
 
-  alias FaktoryWorker.Socket
+  alias FaktoryWorker.Connection
   alias FaktoryWorker.Socket.Tcp
+  alias FaktoryWorker.Protocol
 
-  @spec open(opts :: keyword()) :: {:ok, %Socket{}} | {:error, term()}
+  @enforce_keys [:host, :port, :socket, :socket_handler]
+  defstruct [:host, :port, :socket, :socket_handler]
+
+  @spec open(opts :: keyword()) :: {:ok, %Connection{}} | {:error, term()}
   def open(opts \\ []) do
-    socket_module = Keyword.get(opts, :socket_module, Tcp)
+    socket_handler = Keyword.get(opts, :socket_handler, Tcp)
     host = Keyword.get(opts, :host, "localhost")
     port = Keyword.get(opts, :port, 7419)
 
-    with {:ok, socket} <- socket_module.connect(host, port) do
-      {:ok, socket}
+    with {:ok, connection} <- socket_handler.connect(host, port),
+         :ok <- verify_faktory_version(connection),
+         :ok <- verify_hello_response(connection) do
+      {:ok, connection}
+    end
+  end
+
+  def send_command(%{socket_handler: socket_handler} = connection, command) do
+    case Protocol.encode_command(command) do
+      {:ok, payload} -> socket_handler.send(connection, payload)
+      error -> error
+    end
+  end
+
+  def recv(%{socket_handler: socket_handler} = connection) do
+    socket_handler.recv(connection)
+  end
+
+  defp verify_faktory_version(connection) do
+    case recv(connection) do
+      {:ok, response} ->
+        response
+        |> Protocol.decode_response()
+        |> verify_version()
+
+      error ->
+        error
+    end
+  end
+
+  defp verify_version({:ok, %{"v" => 2}}), do: :ok
+
+  defp verify_version({:ok, %{"v" => version}}) do
+    {:error, "Only Faktory version '2' is supported (connected to Faktory version '#{version}')."}
+  end
+
+  defp verify_hello_response(connection) do
+    with :ok <- send_command(connection, {:hello, 2}),
```

Comment:
I feel like `2` should be extracted to a named module attribute like `@faktory_version`. I wasn't 100% sure what `2` meant here until I checked `Protocol.encode_command/1`.

Reply:
Thanks @beerlington, now updated.